crates/ruff_python_formatter/src/comments/mod.rs

```diff
@@ -1,2 +1,274 @@
 #![allow(unused, unreachable_pub)] // TODO(micha): Remove after using the new comments infrastructure in the formatter.
+
+//! Types for extracting and representing comments of a syntax tree.
+//!
+//! Most programming languages support comments allowing programmers to document their programs.
+//! Comments are different from other syntax because programming languages allow comments in almost any position,
+//! giving programmers great flexibility on where they can write comments:
+//!
+//! ```javascript
+//! /**
+//!  * Documentation comment
+//!  */
+//! async /* comment */ function Test () // line comment
+//! {/*inline*/}
+//! ```
+//!
+//! This flexibility makes formatting comments challenging because:
+//! * The formatter must consistently place comments so that re-formatting the output yields the same result,
+//!   and does not create invalid syntax (line comments).
+//! * It is essential that formatters place comments close to the syntax the programmer intended to document.
+//!   However, the lack of rules regarding where comments are allowed and what syntax they document requires
+//!   the use of heuristics to infer the documented syntax.
+//!
+//! This module tries to strike a balance between placing comments as closely as possible to their source location
+//! and reducing the complexity of formatting comments. It does so by associating comments per node rather than a token.
+//! This greatly reduces the combinations of possible comment positions, but turns out to be, in practice,
+//! sufficiently precise to keep comments close to their source location.
+//!
+//! Luckily, Python doesn't support inline comments, which simplifying the problem significantly.
```

Comment:
Why are you so lucky this time T.T

Reply:
haha.. I'm waiting for the moment where we start formatting other languages and I'll be back in the game